### ISSUE TITLE
EQLHarmonic.grid raises error when passing pyproj projection and extra_coords

### ISSUE THREAD

**Issue by santisoler** [fatiando/harmonica#170]
**Description of the problem**

When trying to grid harmonic data given in Cartesian coordinates through `EQLHarmonic` into a regular grid defined on geodetic, `EQLHarmonic.grid` raises an error due to an extra coordinate being passed as argument to the `pyproj.Proj` instance.

The coordinates generated by the `grid` method are a tuple of `longitude`, `latitude` and `height` (the last one is built after `extra_coords`). They are passed to the passed `pyproj.Proj` instance, but they only expects 2 arguments, instead of 3.

**Full code that generated the error**

For example:
```python
import pyproj
import verde as vd
import harmonica as hm

data = hm.datasets.fetch_south_africa_gravity()

# Choose only a portion of the data points
region_deg = (15, 18, -21, -18)
are_inside = vd.inside((data.longitude, data.latitude), region_deg)
data = data[are_inside]

# Project data points into Cartesian coordinates
projection = pyproj.Proj(proj="merc", lat_ts=(data.latitude.min() + data.latitude.max()) / 2)
easting, northing = projection(data.longitude.values, data.latitude.values)
coordinates = (easting, northing, data.elevation.values)

# Fit equivalent sources
eql = hm.EQLHarmonic(damping=1e1)
eql.fit(coordinates, data.gravity)

# Predict on a regular grid in geodetic coordinates
eql.grid(region=region_deg, spacing=0.1, projection=projection, extra_coords=2000)
```

**Full error message**

```
Traceback (most recent call last):
  File "grid_test.py", line 24, in <module>
    eql.grid(region=region_deg, spacing=0.1, projection=projection, extra_coords=2000)
  File "/home/santi/.miniconda3/envs/harmonica/lib/python3.7/site-packages/verde/base/base_classes.py", line 303, in grid
    data = check_data(self.predict(projection(*coordinates)))
  File "/home/santi/.miniconda3/envs/harmonica/lib/python3.7/site-packages/pyproj/proj.py", line 173, in __call__
    lon, lat = args
ValueError: too many values to unpack (expected 2)
```

**System information**

* Operating system: Manjaro XFCE
* Python installation (Anaconda, system, ETS): Anaconda
* Version of Python: 3.7.6
* Version of this package: 0.1.0+12.g0031491.dirty
* If using conda, paste the output of `conda list` below:

<details>
<summary>output of conda list</summary>
<pre>
# Name                    Version                   Build  Channel
_libgcc_mutex             0.1                 conda_forge    conda-forge
_openmp_mutex             4.5                      1_llvm    conda-forge
alabaster                 0.7.12                     py_0    conda-forge
appdirs                   1.4.3                      py_1    conda-forge
astroid                   2.3.3                    py37_1    conda-forge
attrs                     19.3.0                     py_0    conda-forge
babel                     2.8.0                      py_0    conda-forge
black                     19.10b0                  py37_0    conda-forge
bokeh                     1.4.0                    py37_0    conda-forge
boost-cpp                 1.72.0               h8e57a91_0    conda-forge
boule                     0.1.1                      py_0    conda-forge
bzip2                     1.0.8                h516909a_2    conda-forge
ca-certificates           2020.4.5.1           hecc5488_0    conda-forge
cairo                     1.16.0            hcf35c78_1003    conda-forge
cartopy                   0.18.0           py37h4b180d9_0    conda-forge
certifi                   2020.4.5.1       py37hc8dfbb8_0    conda-forge
cffi                      1.14.0           py37hd463f26_0    conda-forge
cfitsio                   3.470                h3eac812_5    conda-forge
chardet                   3.0.4                 py37_1003    conda-forge
click                     7.0                        py_0    conda-forge
cloudpickle               1.3.0                      py_0    conda-forge
coverage                  5.1              py37h8f50634_0    conda-forge
cryptography              2.8              py37h72c5cf5_1    conda-forge
curl                      7.69.1               h33f0ec9_0    conda-forge
cycler                    0.10.0                     py_2    conda-forge
cytoolz                   0.10.1           py37h516909a_0    conda-forge
dask                      2.11.0                     py_0    conda-forge
dask-core                 2.11.0                     py_0    conda-forge
dbus                      1.13.6               he372182_0    conda-forge
distributed               2.11.0                   py37_0    conda-forge
docutils                  0.16                     py37_0    conda-forge
entrypoints               0.3                   py37_1000    conda-forge
expat                     2.2.9                he1b5a44_2    conda-forge
flake8                    3.7.9            py37hc8dfbb8_1    conda-forge
fontconfig                2.13.1            h86ecdb6_1001    conda-forge
freetype                  2.10.0               he983fc9_1    conda-forge
freexl                    1.0.5             h14c3975_1002    conda-forge
fsspec                    0.6.2                      py_0    conda-forge
gdal                      3.0.4            py37hbb6b9fb_1    conda-forge
geos                      3.8.0                he1b5a44_0    conda-forge
geotiff                   1.5.1                hcbe54f9_9    conda-forge
gettext                   0.19.8.1          hc5be6a0_1002    conda-forge
giflib                    5.2.1                h516909a_2    conda-forge
glib                      2.58.3          py37h6f030ca_1002    conda-forge
gst-plugins-base          1.14.5               h0935bb2_2    conda-forge
gstreamer                 1.14.5               h36ae1b5_2    conda-forge
harmonica                 0.1.0+12.g0031491.dirty           dev_0    <develop>
hdf4                      4.2.13            hf30be14_1003    conda-forge
hdf5                      1.10.5          nompi_h3c11f04_1104    conda-forge
heapdict                  1.0.1                      py_0    conda-forge
icu                       64.2                 he1b5a44_1    conda-forge
idna                      2.9                        py_1    conda-forge
imagesize                 1.2.0                      py_0    conda-forge
importlib_metadata        1.5.0                    py37_0    conda-forge
isort                     4.3.21                   py37_0    conda-forge
jinja2                    2.11.1                     py_0    conda-forge
joblib                    0.14.1                     py_0    conda-forge
jpeg                      9c                h14c3975_1001    conda-forge
json-c                    0.13.1            h14c3975_1001    conda-forge
kealib                    1.4.13               hec59c27_0    conda-forge
kiwisolver                1.1.0            py37hc9558a2_0    conda-forge
krb5                      1.17.1               h2fd8d38_0    conda-forge
lazy-object-proxy         1.4.3            py37h516909a_1    conda-forge
ld_impl_linux-64          2.33.1               h53a641e_8    conda-forge
libblas                   3.8.0               15_openblas    conda-forge
libcblas                  3.8.0               15_openblas    conda-forge
libclang                  9.0.1           default_hde54327_0    conda-forge
libcurl                   7.69.1               hf7181ac_0    conda-forge
libdap4                   3.20.6               h1d1bd15_0    conda-forge
libedit                   3.1.20170329      hf8c457e_1001    conda-forge
libffi                    3.2.1             he1b5a44_1006    conda-forge
libgcc-ng                 9.2.0                h24d8f2e_2    conda-forge
libgdal                   3.0.4                h022d3c0_1    conda-forge
libgfortran-ng            7.3.0                hdf63c60_5    conda-forge
libiconv                  1.15              h516909a_1005    conda-forge
libkml                    1.3.0             hb574062_1011    conda-forge
liblapack                 3.8.0               15_openblas    conda-forge
libllvm8                  8.0.1                hc9558a2_0    conda-forge
libllvm9                  9.0.1                hc9558a2_0    conda-forge
libnetcdf                 4.7.3           nompi_h9f9fd6a_101    conda-forge
libopenblas               0.3.8                h5ec1e0e_0    conda-forge
libpng                    1.6.37               hed695b0_0    conda-forge
libpq                     12.2                 h5513abc_1    conda-forge
libspatialite             4.3.0a            hd318ce7_1036    conda-forge
libssh2                   1.8.2                h22169c7_2    conda-forge
libstdcxx-ng              9.2.0                hdf63c60_2    conda-forge
libtiff                   4.1.0                hc3755c2_3    conda-forge
libuuid                   2.32.1            h14c3975_1000    conda-forge
libwebp                   1.0.2                h56121f0_5    conda-forge
libxcb                    1.13              h14c3975_1002    conda-forge
libxkbcommon              0.10.0               he1b5a44_0    conda-forge
libxml2                   2.9.10               hee79883_0    conda-forge
llvm-openmp               9.0.1                hc9558a2_2    conda-forge
llvmlite                  0.31.0           py37h8b12597_0    conda-forge
locket                    0.2.0                      py_2    conda-forge
lz4-c                     1.8.3             he1b5a44_1001    conda-forge
markupsafe                1.1.1            py37h516909a_0    conda-forge
matplotlib                3.2.0                         1    conda-forge
matplotlib-base           3.2.0            py37h250f245_1    conda-forge
mccabe                    0.6.1                      py_1    conda-forge
more-itertools            8.2.0                      py_0    conda-forge
msgpack-python            1.0.0            py37hc9558a2_0    conda-forge
mypy_extensions           0.4.3                    py37_0    conda-forge
ncurses                   6.1               hf484d3e_1002    conda-forge
nspr                      4.25                 he1b5a44_0    conda-forge
nss                       3.47                 he751ad9_0    conda-forge
numba                     0.48.0           py37hb3f55d8_0    conda-forge
numpy                     1.18.4           py37h8960a57_0    conda-forge
numpydoc                  0.9.2                      py_0    conda-forge
olefile                   0.46                       py_0    conda-forge
openjpeg                  2.3.1                h981e76c_3    conda-forge
openssl                   1.1.1g               h516909a_0    conda-forge
owslib                    0.19.1                     py_0    conda-forge
packaging                 20.1                       py_0    conda-forge
pandas                    1.0.3            py37h0da4684_1    conda-forge
partd                     1.1.0                      py_0    conda-forge
pathspec                  0.7.0                      py_0    conda-forge
pcre                      8.44                 he1b5a44_0    conda-forge
pillow                    7.0.0            py37hefe7db6_0    conda-forge
pip                       20.1               pyh9f0ad1d_0    conda-forge
pixman                    0.38.0            h516909a_1003    conda-forge
pluggy                    0.13.0                   py37_0    conda-forge
pooch                     1.1.0                      py_0    conda-forge
poppler                   0.67.0               h14e79db_8    conda-forge
poppler-data              0.4.9                         1    conda-forge
postgresql                12.2                 h8573dbc_1    conda-forge
proj                      6.3.1                hc80f0dc_1    conda-forge
psutil                    5.7.0            py37h516909a_0    conda-forge
pthread-stubs             0.4               h14c3975_1001    conda-forge
py                        1.8.1                      py_0    conda-forge
pycodestyle               2.5.0                      py_0    conda-forge
pycparser                 2.20                       py_0    conda-forge
pyepsg                    0.4.0                      py_0    conda-forge
pyflakes                  2.1.1                      py_0    conda-forge
pygments                  2.5.2                      py_0    conda-forge
pykdtree                  1.3.1           py37hc1659b7_1002    conda-forge
pylint                    2.4.4                    py37_0    conda-forge
pyopenssl                 19.1.0                     py_1    conda-forge
pyparsing                 2.4.6                      py_0    conda-forge
pyproj                    2.6.0            py37heba2c01_0    conda-forge
pyqt                      5.12.3           py37hcca6a23_1    conda-forge
pyqt5-sip                 4.19.18                  pypi_0    pypi
pyqtwebengine             5.12.1                   pypi_0    pypi
pyshp                     2.1.0                      py_0    conda-forge
pysocks                   1.7.1                    py37_0    conda-forge
pytest                    5.4.1            py37hc8dfbb8_0    conda-forge
pytest-cov                2.8.1                      py_0    conda-forge
python                    3.7.6           h8356626_5_cpython    conda-forge
python-dateutil           2.8.1                      py_0    conda-forge
python_abi                3.7                     1_cp37m    conda-forge
pytz                      2019.3                     py_0    conda-forge
pyyaml                    5.3              py37h516909a_0    conda-forge
qt                        5.12.5               hd8c4c69_1    conda-forge
readline                  8.0                  hf8c457e_0    conda-forge
regex                     2020.2.20        py37h516909a_0    conda-forge
requests                  2.23.0                   py37_0    conda-forge
scikit-learn              0.22.2.post1     py37hcdab131_0    conda-forge
scipy                     1.4.1            py37ha3d9a3c_3    conda-forge
setuptools                45.2.0                   py37_0    conda-forge
shapely                   1.7.0            py37h5d51c17_0    conda-forge
six                       1.14.0                   py37_0    conda-forge
snowballstemmer           2.0.0                      py_0    conda-forge
sortedcontainers          2.1.0                      py_0    conda-forge
sphinx                    2.2.1                      py_0    conda-forge
sphinx-gallery            0.6.2                      py_0    conda-forge
sphinx_rtd_theme          0.4.3                      py_0    conda-forge
sphinxcontrib-applehelp   1.0.2                      py_0    conda-forge
sphinxcontrib-devhelp     1.0.2                      py_0    conda-forge
sphinxcontrib-htmlhelp    1.0.3                      py_0    conda-forge
sphinxcontrib-jsmath      1.0.1                      py_0    conda-forge
sphinxcontrib-qthelp      1.0.3                      py_0    conda-forge
sphinxcontrib-serializinghtml 1.1.4                      py_0    conda-forge
sqlite                    3.30.1               hcee41ef_0    conda-forge
tbb                       2018.0.5             h2d50403_0    conda-forge
tblib                     1.6.0                      py_0    conda-forge
tiledb                    1.7.0                hcde45ca_2    conda-forge
tk                        8.6.10               hed695b0_0    conda-forge
toml                      0.10.0                     py_0    conda-forge
toolz                     0.10.0                     py_0    conda-forge
tornado                   6.0.4            py37h516909a_0    conda-forge
typed-ast                 1.4.1            py37h516909a_0    conda-forge
typing_extensions         3.7.4.1                  py37_0    conda-forge
tzcode                    2020a                h516909a_0    conda-forge
urllib3                   1.25.7                   py37_0    conda-forge
verde                     1.4.0                      py_0    conda-forge
wcwidth                   0.1.8                      py_0    conda-forge
wheel                     0.34.2                     py_1    conda-forge
wrapt                     1.12.0           py37h516909a_0    conda-forge
xarray                    0.15.1                     py_0    conda-forge
xerces-c                  3.2.2             h8412b87_1004    conda-forge
xorg-kbproto              1.0.7             h14c3975_1002    conda-forge
xorg-libice               1.0.10               h516909a_0    conda-forge
xorg-libsm                1.2.3             h84519dc_1000    conda-forge
xorg-libx11               1.6.9                h516909a_0    conda-forge
xorg-libxau               1.0.9                h14c3975_0    conda-forge
xorg-libxdmcp             1.1.3                h516909a_0    conda-forge
xorg-libxext              1.3.4                h516909a_0    conda-forge
xorg-libxrender           0.9.10            h516909a_1002    conda-forge
xorg-renderproto          0.11.1            h14c3975_1002    conda-forge
xorg-xextproto            7.3.0             h14c3975_1002    conda-forge
xorg-xproto               7.0.31            h14c3975_1007    conda-forge
xz                        5.2.4             h14c3975_1001    conda-forge
yaml                      0.2.2                h516909a_1    conda-forge
zict                      2.0.0                      py_0    conda-forge
zipp                      3.1.0                      py_0    conda-forge
zlib                      1.2.11            h516909a_1006    conda-forge
zstd                      1.4.4                h3b9ef0a_1    conda-forge
</pre>
</details>

**Comment by leouieda** [comment]
Ah, this is tricky. One workaround now is to wrap the pyproj function: `lambda lon, lat, h: projection(lon, lat)`. The other way would be to change the `grid` code but then it wouldn't work for a 3D projection function like `WGS84.spherical_to_geodetic` for example.

**Comment by santisoler** [comment]
Yes... this is tricky. Wrapping the pyproj function won't work because the output of `lambda lon, lat, h: projection(lon, lat)` would be `easting, northing` but we will actually need the height for predicting.
And we definitely should make this compatible both with projections and coordinates conversions.

I don't know if there's a way to automatically check if the passed `projection` needs 2 or 3 positional arguments, although even if it exists I don't know if that's a good idea.
Other option would be to add another optional argument for specifying if the `projection` works only with `lon, lat` of with `lon, lat, height`.

**Comment by leouieda** [comment]
> Wrapping the pyproj function won't work

That's a bug in my example. We could do 
 
```python
def proj(lon, lat, h):
    x, y = projection(lon, lat)
    return x, y, h
```

But that is tiresome and I wouldn't want to have that be the only option.

> I don't know if there's a way to automatically check if the passed projection needs 2 or 3 positional arguments,

Yes there is: https://docs.python.org/3/library/inspect.html#introspecting-callables-with-the-signature-object

> Other option would be to add another optional argument for specifying if the projection works only with lon, lat of with lon, lat, height.

Right, so this would require:

* A bug fix in verde to only pass in the first two coordinates to projection 
* Adding a new argument and `if` block to use it. 

It would be great if we could figure out a way of not doing this, though. It will be a frequent source of confusion regarding which argument to use.

**Comment by santisoler** [comment]
This issue was fixed by fatiando/verde#264.
We should pin Verde to v1.5.0.
